docs/mkvtool.1.md
# NAME

mkvtool

# SYNOPSIS

  **mkvtool [global options] command [command-options] [ARGs...]**

# DESCRIPTION

A swiss army knife for easy operations on Matroska containers.

This program provides a simpler front-end to common operations involving
Matroska containers. It hides details and complexities from utilities in the
mkvtoolnix suite, such as mkvmerge, mkvpropedit, and others.

# GLOBAL OPTIONS

* **--help**: Show context-sensitive help.

* **-n**, **--dry-run**: Dry-run mode (only show commands or output.)

* **-q**, **--quiet**: Reduce output, do not show results after merge and remux commands.

* **-J**, **--json**: Output information in JSON format (as an array).

# COMMANDS

Some commands have their own options, in addition to global options. Please
keep in mind that global options must appear before the command and command
options must appear after the command, and before any non-option (like bare
filenames.)

## **help [COMMAND]**

Show a simple help page.

## **merge options [INPUT_FILEs]...**

Merge multiple input files (containing their respective media tracks) into
`<output-file>`.

This operation takes multiple input files (of any format recognized by
mkvmerge) and creates a new MKV file containing all tracks read from the input
files. A common use of this feature is to add a subtitle track to an existing
file that contains only A/V tracks (E.g., to create a MKV file with the A/V
tracks from an MP4 file and subtitles from a SRT file.)

After a successful merge, `mkvtool` will show the tracks in the new file, unless
the global `--quiet` flag is used.

**Options**:

* **--subs**: Copy all subtitles in `<input-file>` (default=true). Use
  `--nosubs` to replace all input subtitle tracks in a MKV file with one or
  more subtitle tracks coming from external files (E.g, SRT files).

  Note: The program does not check the contents of the input files. It is
  entirely possible to create files with multiple video, audio, and subtitle
  tracks (although we typically want only one video track in the output file.)

*  **-o OUTPUT_FILE, --output=OUTPUT_FILE**: Output file.

## **remux [options] INPUT_FILE OUTPUT_FILE**

Remux the input file into the output file, with the option to filter tracks.

After a successful remux, `mkvtool` will show the tracks in the new file, unless
the global `--quiet` flag is used.

**Options**:

* **--tracks**: Command separated list of raw track numbers to be copied. Use `mkvtool show` to
  see the full list of tracks with numbers. This option is mutually exclusive with `--lang`.

* **--lang**: Command separated list of track types/language specifiers to be
  copied.  The format the specifier is `[type]:language`, where type is "v" for
  video, "a" for audio, or "s" for subtitles and "language" is the three letter
  language code. The "language" field can also be **all**, meaning "copy all
  tracks of this type", **none**, for "copy no tracks of this type", and
  **first**, meaning "Only copy the first track of this type".

  If the lang specifier starts with a "!" the track will be excluded.

  By default, all tracks are copied (v:all,a:all,s:all), so you don't need to
  start the track list with "all".

* **--ignore**: Tracks matching the string specified by this option will not be
  copied to the output.  Useful to prevent the copy of Forced or SDH subtitles
  for example.

**Examples:**

* Copy all video tracks, English audio, and all subtitle tracks from in.mkv to
  out.mkv.  Note how by default, all tracks are copied so we don't need to use
  `v:all` and `a:all`:

```
mkvtool remux --lang='a:eng' in.mkv out.mkv
```

* Copy all video tracks, all audio tracks except Russian and French, and no subtitle tracks:

```
mkvtool remux --lang='!a:rus,!a:fre,s:none'
```

* Copy all tracks, but exclude any tracks with "Forced" in the description (as in "forced subtitles"):

```
mkvtool remux --ignore="Forced" in.mkv out.mkv
```

* Copy only all video tracks, the first audio track, and all subtitles in the English language:

```
mkvtool remux --lang='a:first,s:eng'
```

## **rename INPUT_FILEs...**

Rename `<input-files>` into a standardized format, using metadata in the
original filenames.

Attempt to fetch metadata (title, season, year, resolution, etc) from each of
the filenames specified in `<input-files>`. The filenames themselves are parsed
for data, so this is a "best effort" operation.  Each filename is renamed
following a simplified format, compatible with most streaming applications.

Note: For TV series, There's no clean way to fetch the episode name, so this
information will be lost during the rename. Most streaming servers fill that
information in their databases based on Title, Episode, and Season, so that
tends not to be a problem for most people.

## **setdefault TRACK FILEs...**

Set the track specified with the `<track>` argument as the default track
in `<mkvfile>`.

Default tracks tell the player which track (in the presence of multiple tracks
of the same type) should be played. It's most commonly used for audio and
subtitle tracks. Not all players obey these settings (bust empirically, most
appear to do the right thing.)

## **setdefaultbylang [options] FILEs...**

Set the track with the first matching language as the default track.

* **--lang=LANGUAGE**: languages to match. It's possible to specify `--lang` multiple
  times, in which case the first match will prevail.

There's also the "default" meta-language, which matches the default
language in the MKV file (normally, shown as an empty language when
using the `show` option).

For example, using `--lang=en --lang=es --lang=default` will cause the program
to first attempt to find a subtitle track in the English language. If no
subtitle tracks in English exist, it will attempt Spanish next. As a last
resort, a subtitle track with the "default" language will be matched and set as
the default subtitle track (if it exists.)

* **--ignore=STRING**: Ignore any tracks containing the specified string in the
  description.  This is useful to ignore "Forced" subtitles, for example.

This command only works in subtitle tracks for now.

**Example:**

This will set the first subtitle track in English (change to your favorite
language) to be the default subtitle track. Failing that, it will attempt the
default language and "und" (some files have one subtitle track marked as
"undefined".)

```
mkvtool setdefaultbylang --lang=eng --lang=default --lang=und --ignore="force" *.mkv
```


## **show [options] INPUT_FILEs...**

Shows a listing of all tracks in the file.

Options:

* **-u, --uid**: Include track UIDs in the output (table output only).

## **version**

Show version information.

# Author

- (C) 2021-2026 by Marco Paganini <paganini at paganini dot net>



mkvtool.go
// This file is part of mkvtool (http://github.com/marcopaganini/mkvtool))
// See instructions in the README.md file that accompanies this program.
// (C) 2022-2024 by Marco Paganini <paganini AT paganini DOT net>

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/fatih/structs"
	"github.com/jedib0t/go-pretty/table"
	ParseTorrentName "github.com/middelink/go-parse-torrent-name"
	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)

// A friendly chat about Matroska metadata track numbers.
//
// Matroska tracks numbers are confusing. Tracks are stored in the file
// starting at 1 (ONE). Some mkvtoolnix commands such as mkvmerge and
// mkvextract expect tracks to start at offset zero (ZERO), while others like
// mkvpropedit, expect offset 1. Due to this, the following conventions were
// adopted here:
//
// - Tracks are always displayed starting at 0 (as the output of mkvmerge --identify)
// - Any actions using mkvpropedit automatically add one to the track number.
// - Any actions using mkvmerge or mkvextract will use the track number unchanged.
//
// Track Types. See https://www.matroska.org/technical/specs/index.html
const (
	typeAudio     = "audio"
	typeVideo     = "video"
	typeSubtitles = "subtitles"
)

// BuildVersion holds the git build number (set by make).
var BuildVersion string

// showJSON displays the track information in JSON format as an array.
func showJSON(mkv []matroska) error {
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	return enc.Encode(mkv)
}

// show lists all tracks in a file.
var show = func(mkv matroska, showUID bool) {
	tab := table.NewWriter()
	tab.SetOutputMirror(os.Stdout)
	if showUID {
		tab.AppendHeader(table.Row{"Number", "UID", "Type", "Name", "Language", "Codec", "Default"})
	} else {
		tab.AppendHeader(table.Row{"Number", "Type", "Name", "Language", "Codec", "Default"})
	}

	for _, track := range mkv.Tracks {
		// Create a row with the desired columns.
		// mkvmerge reports tracks starting at zero, so we add one to match the file.
		row := []interface{}{track.ID}
		if showUID {
			row = append(row, uint64(track.Properties.UID))
		}
		row = append(row, track.Type, track.Properties.TrackName, track.Properties.Language, track.Codec)

		// Make default flag easier to see.
		if track.Properties.DefaultTrack {
			row = append(row, "<=====")
		} else {
			row = append(row, "")
		}
		tab.AppendRow(row)
	}
	tab.Render()
}

// setdefault resets flagDefault on all subtitle tracks and sets it on the chosen track UID.
var setdefault = func(mkv matroska, tracknum int, cmd runner) error {
	command := []string{
		"mkvpropedit",
		mkv.FileName,
	}

	for _, track := range mkv.Tracks {
		if track.Type == typeSubtitles {
			// mkvpropedit uses base 1 for track (not zero).
			command = append(command, "--edit", fmt.Sprintf("track:%d", track.ID+1), "--set", "flag-default=0")
		}
	}

	if err := cmd.run(command[0], command[1:]...); err != nil {
		return err
	}
	return adddefault(mkv, tracknum, cmd)
}

// trackByLanguageAndType returns the track number (base 0) for the first track
// with one of the specified languages matching the track type. The list of
// languages works as a priority, meaning that languages=["eng","fra"] will
// first attempt to find a track with the English language, and failing that,
// French. The special language "default" will cause tracks without a language
// code to be selected (Matroska has the concept of a "default language", which
// is usually English -- tracks with this language will not have a language
// code).
//
// The ignore slice contains a list of strings for case-insentive search
// against the track name. If the selected language contains one of the strings
// in this slice, it will be ignored. This is useful to select tracks by
// language while ignoring 'Forced' tracks.
var trackByLanguageAndType = func(mkv matroska, languages []string, tracktype string, ignore []string) (int, error) {
	for _, lang := range languages {
		if lang == "default" {
			lang = ""
		}
		for _, track := range mkv.Tracks {
			// Match subtitle and language.
			if track.Type != tracktype || track.Properties.Language != lang {
				continue
			}
			// Make sure track should not be ignored.
			if stringInSlice(track.Properties.TrackName, ignore) {
				continue
			}
			return track.ID, nil
		}
	}
	return 0, fmt.Errorf("no track with language(s): %s", strings.Join(languages, ","))
}

// stringInSlice returns true if a string exists inside a slice of strings.
// Comparison is case insensitive.
func stringInSlice(s string, slc []string) bool {
	for _, substr := range slc {
		if strings.Contains(strings.ToLower(s), strings.ToLower(substr)) {
			return true
		}
	}
	return false
}

// remux re-multiplexes the input file(s) into the output file. Setting subs to
// false will cause subs not to be copied.
func remux(infiles []string, outfile string, cmd runner, subs bool) error {
	cmdline := []string{"mkvmerge"}
	if !subs {
		cmdline = append(cmdline, "-S")
	}
	cmdline = append(cmdline, infiles...)
	cmdline = append(cmdline, "-o", outfile)

	return cmd.run(cmdline[0], cmdline[1:]...)
}

// adddefault adds the default flag to a given track UID.
func adddefault(mkv matroska, tracknum int, cmd runner) error {
	for _, track := range mkv.Tracks {
		if track.ID == tracknum {
			// mkvpropedit uses base 1 for tracks.
			return cmd.run("mkvpropedit", mkv.FileName, "--edit", fmt.Sprintf("track:%d", tracknum+1), "--set", "flag-default=1")
		}
	}
	return fmt.Errorf("file %s does not contain track %d", mkv.FileName, tracknum)
}

// rename renames a file according to the "Scene" information in the file.
var rename = func(mask, fname string, dryrun bool) error {
	newname, err := format(fname, mask)
	if err != nil {
		return err
	}
	dir, _ := filepath.Split(fname)
	newfile := filepath.Join(dir, newname)

	fmt.Printf("%s => %s\n", fname, newfile)
	if dryrun {
		return nil
	}
	return os.Rename(fname, newfile)
}

// format parses "Scene" information in the file and returns a string formatted
// according to a formatting mask. The mask may contain the following tokens:
//
// %[format]{audio}
// %[format]{codec}
// %[format]{container} (this matches the original extension)
// %[format]{episode}
// %[format]{episodeName}
// %[format]{excess}
// %[format]{extended}
// %[format]{garbage}
// %[format]{group}
// %[format]{hardcoded}
// %[format]{language}
// %[format]{proper}
// %[format]{quality}
// %[format]{region}
// %[format]{repack}
// %[format]{resolution}
// %[format]{season}
// %[format]{title}
// %[format]{website}
// %[format]{widescreen}
// %[format]{year}
//
// Where "format" is a printf style format sizing specification. Complete
// examples:
// - %-02.2{season} - Season formatted as two characters, left padded wth zeroes.
// - %-20{title} - Title truncated to 20 characters
//
// Anything not matching the %[format]{xxxx} construct will be interpreted literally.
//
// Formatting will fail if any element present in the mask cannot be resolved
// (a typical example is asking for episode numbers for movies).
var format = func(mask, fname string) (string, error) {
	// Split the filename so we can work on parts separately.
	_, file := filepath.Split(fname)

	parsed, err := ParseTorrentName.Parse(file)
	if err != nil {
		return "", err
	}
	fields := structs.Map(parsed)

	// tags are formatted as %[format]{value}
	re, err := regexp.Compile(`%((?:-?[\d]+)?(?:\.\d+)?){([a-z]+)}`)
	if err != nil {
		return "", err
	}

	var errlist []string

	formatted := re.ReplaceAllStringFunc(mask, func(match string) string {
		// Split matched tag into size formatting specifier and tag name.
		// Tag must be capitalized to match the keys in the map.
		e := re.FindStringSubmatch(match)
		sizespec := e[1]
		tag := cases.Title(language.English).String(e[2])

		if i, ok := fields[tag]; ok {
			switch t := i.(type) {
			case string:
				val := i.(string)
				if val == "" {
					break
				}
				// Special case for title: Capitalize
				if tag == "Title" {
					val = cases.Title(language.English).String(val)
				}
				return fmt.Sprintf("%"+sizespec+"s", val)
			case int:
				val := i.(int)
				if val <= 0 {
					break
				}
				return fmt.Sprintf("%"+sizespec+"d", i.(int))
			default:
				errlist = append(errlist, fmt.Sprintf("Internal error: Unknown type %T for %q", match, t))
				return "*ERROR*"
			}
		}
		errlist = append(errlist, fmt.Sprintf("Unable to parse data for %s", match))
		return "*ERROR*"
	})

	if len(errlist) != 0 {
		return "", fmt.Errorf("%s", strings.Join(errlist, ";"))
	}
	return formatted, nil
}

// requirements returns nil if all required tools are installed and an error indicating
// the tools missing otherwise.
func requirements() error {
	var tools = []string{"mkvextract", "mkvmerge", "mkvpropedit"}

	missing := []string{}
	for _, t := range tools {
		_, err := exec.LookPath(t)
		if err != nil {
			missing = append(missing, t)
		}
	}
	if len(missing) != 0 {
		return fmt.Errorf("required 3rd party tool(s) missing: %s", strings.Join(missing, ","))
	}
	return nil
}

// mustParseFile parses the MKV file using the JSON output from mkmerge --identify.
// error message in case of problems.
var mustParseFile = func(fname string) matroska {
	var stdout bytes.Buffer

	cmd := exec.Command("mkvmerge", "--identify", "-F", "json", fname)
	cmd.Stdout = &stdout
	err := cmd.Run()
	if err != nil {
		log.Printf("mkvmerge failed: %v", err)
		log.Fatalf("--- Output ---\n%s\n", stdout.String())
	}

	// Decode JSON.
	var mkv matroska
	err = json.Unmarshal(stdout.Bytes(), &mkv)
	if err != nil {
		log.Fatalf("Error decoding JSON output from mkvmerge: %v", err)
	}
	return mkv
}
